### PR TITLE
Migration Guide for Flutter web slotted rendering.

### PR DESCRIPTION
**Pull request by ditman**
This change adds a Migration Guide for the new Flutter web slotted rendering, explaining how the code is changing and what users need to do for their web apps to keep working as normal.

Fixes: https://github.com/flutter/flutter/issues/80524

### CODE REVIEW

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
+---
+title: Using HTML slots to render platform views in the web
+description: iframes in Flutter web used to reload, because of the way some DOM operations were made. A change in the way Flutter web apps render Platform Views makes them much more stable (preventing iframe reloads, and other problems with video tags or forms potentially losing their state).
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:3]: Remove "much more". It wasn't stable and now it simply is :)

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+---
+
+## Summary
+
+Flutter now renders Platform Views using [slot elements][] inside of a single,
+app-wide [Shadow Root][]. Slot elements can be added/removed/moved around the
+Shadow DOM without affecting the underlying slotted content (which is rendered
+in a constant location)
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:11]: This discusses an implementation detail, which may be useful in the "Context" section below. In the summary I think the developers would be most interested in what happens to the HTML DOM that _they_ created, so I'd suggest that we talk about how we put their platform views into the root element, instead of injecting it into the Flutter's internal DOM structures.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+## Context
+
+The Flutter framework frequently tweaks its Render Tree to optimize the paint
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:15]: I'm not sure if "render tree" is worth capitalizing.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+Using certain DOM elements as the "target" of some DOM operations causes those
+elements to lose their internal state. In practice this means that `iframe` tags
+are going to reload, `video` players might restart, or a form being edited might
+lose the edits made to it.
+
+This change was made to:
+
+* make Platform Views behave in a stable way in Flutter web.
+* unify the way Platform Views were rendered in the web for both rendering
+backends (`html` and `canvaskit`).
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:31]: And:

* provide a predictable location in the DOM to allow developer to reliably use CSS to style HTML inside platform views, and use other standard DOM API, such as `querySelector`, and `getElementById`.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+Shadow DOM without affecting the underlying slotted content (which is rendered
+in a constant location)
+
+## Context
+
+The Flutter framework frequently tweaks its Render Tree to optimize the paint
+operations that are ultimately made per frame. In the web, these Render Tree
+changes often result in DOM operations.
+
+Flutter web used to render its Platform Views (`HtmlElementView` widgets)
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:19]: I'd probably also not capitalize "platform views"

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+This change was made to:
+
+* make Platform Views behave in a stable way in Flutter web.
+* unify the way Platform Views were rendered in the web for both rendering
+backends (`html` and `canvaskit`).
+
+## Description of change
+
+A Flutter web app is now rendered inside a common Shadow Root in which slot
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:35]: I'd not capitalize "shadow root", but I would link to MDN or some other reliable source that explains what it is.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+elements are used to represent platform views. The actual content of each
+platform view is rendered *outside* of the Shadow Root, as a child of the Shadow
+Root host.
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:38]: I think we can gzip this sentence a bit: "...is rendered as a sibling of the shadow root"

Comment (yjbanov) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:38]: Or, since we're keeping `<flt-glass-pane>` we could just say "direct child of `<flt-glass-pane>`"

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+From an app's perspective, this change is transparent, **however** this was
+deemed a _breaking change_ because tests that make assumptions about the DOM of
+a Flutter web app can break.
+
+## Migration guide
+
+### Code
+
+The content returned by the users' Platform View Factory used to be resized and
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:95]: nit: user's

Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:95]: nit: s/resized/sized/

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+positioned by the framework, but not anymore. Ensure that your returned
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:96]: I'd insert another sentence between "but not anymore." and "Ensure...":

...but not anymore. Instead, Flutter sizes and positions `<flt-platform-view>`, which is the parent of the content. Ensure...

Comment (yjbanov) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:96]: I'd also replace "Ensure" with "To ensure". Maybe the developer doesn't want to fill the whole space. They only need to set size to 100% if they want to.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+  return htmlElement;
+});
+```
+
+Otherwise the engine will print a warning message to the console similar to:
+
+<!-- skip -->
+```bash
+Height of Platform View type: [$viewType] may not be set. Defaulting to `height: 100%`.
+Set `style.height` to any appropriate value to stop this message.
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:117]: Hmm, do we need this? Why can't I use relative layout (flexbox, etc) to fill the space?

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+```
+
+or:
+
+<!-- skip -->
+```bash
+Width of Platform View type: [$viewType] may not be set. Defaulting to `width: 100%`.
+Set `style.width` to any appropriate value to stop this message.
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:125]: same here

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+```
+
+### Tests
+
+Unfortunately, there are no selectors that let one _pierce_ through the Shadow
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:130]: s/Unfortunately/By design/ (this is why we use it 😈)

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+DOM boundary to locate elements underneath it, so code that once peeked inside
+of the `flt-glass-pane`, needs to be made Shadow DOM aware. Using the html
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:132]: nit: remove comma before "needs"

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+// Find #some-element inside the app:
+final element = document.querySelector('#some-element');
+```
+
+Code after migration:
+
+<!-- skip -->
+```dart
+final shadowRoot = document.querySelector('flt-glass-pane')?.shadowRoot!;
+final element = shadowRoot.querySelector('#some-element');
```
Comment (yjbanov) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:146]: Why is this necessary? After the migration `#some-element` will no longer be inside the shadow DOM, only the slot. We don't want people looking into Flutter's private DOM structure. It's unpredictable and unstable.

Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:146]: This is what most of the tests that failed were doing. We can document the fact that now they shouldn't drill inside the shadowDOM, as mentioned above, that's probably better!

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 10 above)
Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:117]: If you use any other way of filling the space, you need to set height/width to `auto`, or anything that is not empty. In the engine the only thing we look at is that width/height is set.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 1 above)
Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:3]: Hah, yep. Removed!

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 7 above)
Comment (sfshaza2) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:38]: I'm suggesting a change that corresponds to Yegor's comment However, he might have intended for that second sentence to be deleted? @yjbanov?

```suggestion
A Flutter web app is now rendered inside a common [shadow root][] in which slot
elements represent platform views. The actual content of each
platform view is rendered *outside* of the shadow root, as a child of the shadow
root host.

[shadow root]: https://www.html5rocks.com/en/tutorials/webcomponents/shadowdom-201/
```

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+  ...
+</flt-glass-pane>
+
+...
+```
+
+Now, when the framework needs to move DOM nodes around, it'll operate over
+`flt-platform-view-slot`s, which only contain the `slot` tag that "projects" the
+content defined in the `flt-platform-view` tags outside of the Shadow Root
+(which never move), thus preventing the reloads.
```
Comment (sfshaza2) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:85]:
```suggestion
After this change, when the framework needs to move DOM nodes around,
it operates over`flt-platform-view-slot`s, which only contain the `slot` tag that "projects" the
content defined in the `flt-platform-view` tags outside of the shadow root
(which never move), thus preventing the reloads.
```

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 3 above)
Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:15]: Capitalization Of Things, The Bane Of My Existence :P

Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:15]: (lower-cased :P)

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+From an app's perspective, this change is transparent, **however** this was
+deemed a _breaking change_ because tests that make assumptions about the DOM of
+a Flutter web app can break.
```
Comment (sfshaza2) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:89]:
```suggestion
From an app's perspective, this change is transparent. **However**, this is
considered a _breaking change_ because some tests make assumptions
about how the DOM of a Flutter web app might break.
```

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 2 above)
Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:11]: I moved this to the `context` section.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+
+## Migration guide
+
+### Code
+
+The content returned by the users' Platform View Factory used to be resized and
+positioned by the framework, but not anymore. Ensure that your returned
+`html.Element` takes the whole space allocated for it by the framework, by
+setting its `style.width` and `style.height` properties to `'100%'`:
```
Comment (sfshaza2) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:98]:
```suggestion
Previously, the content returned by [`PlatformViewFactory`][] was resized and
positioned by the framework. Instead, Flutter now sizes and positions
`<flt-platform-view>`, which is the parent of the content.
To ensure the `html.Element` that you return takes the whole space
allocated to it, set its `style.width` and `style.height` properties to `'100%'`:

[`PlatformViewFactory`]: {{site.api}}/javadoc/index.html?io/flutter/plugin/platform/PlatformView.html
```

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+  final html.Element htmlElement = html.DivElement()
+    // ..other props
+    ..style.width = '100%'
+    ..style.height = '100%';
+  // ...
+  return htmlElement;
+});
+```
+
+Otherwise the engine will print a warning message to the console similar to:
```
Comment (sfshaza2) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:112]:
```suggestion
Otherwise, the engine prints a warning message to the console similar to:
```

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 13 above)
Comment (sfshaza2) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:132]:
```suggestion
By design, there are no selectors that let one _pierce_ through the shadow
DOM boundary to locate elements underneath it, so code that once peeked inside
of the `flt-glass-pane` needs to be made shadow DOM aware. Using the html
```

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+above).
+
+## Timeline
+
+Landed in version: 2.3.0-16.0.pre<br>
+In stable release: not yet
+
+## References
+
+Design Doc:
```
Comment (sfshaza2) [inline src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:174]:
```suggestion
Design doc:
```

**Review by sfshaza2 — CHANGES_REQUESTED** [review]
Yegor made some comments that I turned into official GitHub suggestions. He also had some questions as to the design, and I will leave that to the two of you to resolve.  :)

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 10 above)
Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:117]: I'll add a "But I don't want to set width: 100% in my platform view" section

**Comment by ditman** [comment]
Thanks @sfshaza2 for the suggestions! I was halfway through applying by hand Yegor's suggestions so I took the ones from you that wouldn't directly conflict with my changes :P I've manually applied your changes to the first half of the doc, I think I got'em'all!

**Review by sfshaza2 — APPROVED** [review]
LGTM, but I will let Yegor have the final say.

File: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md (hunk 14 above)
Comment (ditman) [reply src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:162]: Removed this, warned people against looking into the shadowDom of the app.

**Comment by ditman** [comment]
I've added a few more lines here and there, I think I addressed all of the comments. @yjbanov I've re-tagged you to PTAL. Once we're happy with the content, I'll ping @sfshaza2 for a last review!

**Review by yjbanov — APPROVED** [review]
lgtm
